api4/config_test.go

```diff
@@ -342,3 +342,86 @@ func TestGetOldClientConfig(t *testing.T) {
 		}
 	})
 }
+
+func TestPatchConfig(t *testing.T) {
+	th := Setup().InitBasic()
+	defer th.TearDown()
+	client := th.Client
+
+	t.Run("config is missing", func(t *testing.T) {
+		_, response := client.PatchConfig(nil)
+		CheckBadRequestStatus(t, response)
+	})
+
+	t.Run("user is not system admin", func(t *testing.T) {
+		_, response := client.PatchConfig(&model.Config{})
+		CheckForbiddenStatus(t, response)
+	})
+
+	t.Run("should not update the restricted fields when restrict toggle is on", func(t *testing.T) {
+		*th.App.Config().ExperimentalSettings.RestrictSystemAdmin = true
+
+		config := model.Config{LogSettings: model.LogSettings{
+			ConsoleLevel: model.NewString("INFO"),
+		}}
+
+		updatedConfig, _ := th.SystemAdminClient.PatchConfig(&config)
+
+		assert.Equal(t, "DEBUG", *updatedConfig.LogSettings.ConsoleLevel)
+	})
+
+	t.Run("check if config is valid", func(t *testing.T) {
+		config := model.Config{PasswordSettings: model.PasswordSettings{
+			MinimumLength: model.NewInt(4),
+		}}
+
+		_, response := th.SystemAdminClient.PatchConfig(&config)
+
+		assert.Equal(t, http.StatusBadRequest, response.StatusCode)
+		assert.NotNil(t, response.Error)
+		assert.Equal(t, "model.config.is_valid.password_length.app_error", response.Error.Id)
+	})
+
+	t.Run("should patch the config", func(t *testing.T) {
+		*th.App.Config().ExperimentalSettings.RestrictSystemAdmin = false
+
+		config := model.Config{PasswordSettings: model.PasswordSettings{
+			Lowercase: model.NewBool(true),
+			Number:    model.NewBool(true),
+			Uppercase: model.NewBool(true),
+			Symbol:    model.NewBool(true),
+		}, LogSettings: model.LogSettings{
+			ConsoleLevel: model.NewString("INFO"),
+		}}
```

Comment:
Just wondering if we could be explicit about:
1. Testing a before and after the patch
2. Patching a few different structs in the config
3. Maybe a couple that are nested a level or two
4. And maybe a couple of different types (string, int, slice)

Reply:
Fair point. I have made changes. Before patching the config I make sure what are the values in config then I call patch and once again at the end make sure the values have changed. I have added added some different structs. Please have a look.